server/src/internalClusterTest/java/org/opensearch/wlm/WorkloadManagementIT.java

```diff
@@ -0,0 +1,403 @@
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(type);
+            out.writeString(queryGroupId);
+        }
+
+        public String getType() {
+            return type;
+        }
+
+        public String getQueryGroupId() {
+            return queryGroupId;
+        }
+    }
+
+    public static class TestQueryGroupTaskTransportAction extends HandledTransportAction<TestQueryGroupTaskRequest, TestResponse> {
+        public static final ActionType<TestResponse> ACTION = new ActionType<>("internal::test_query_group_task_action", TestResponse::new);
+        private final ThreadPool threadPool;
+
+        @Inject
+        public TestQueryGroupTaskTransportAction(TransportService transportService, ThreadPool threadPool, ActionFilters actionFilters) {
+            super(ACTION.name(), transportService, actionFilters, TestQueryGroupTaskRequest::new);
+            this.threadPool = threadPool;
+        }
+
+        @Override
+        protected void doExecute(Task task, TestQueryGroupTaskRequest request, ActionListener<TestResponse> listener) {
+            threadPool.getThreadContext().putHeader(QUERY_GROUP_ID_HEADER, request.getQueryGroupId());
+            threadPool.executor(ThreadPool.Names.SEARCH).execute(() -> {
+                try {
+                    CancellableTask cancellableTask = (CancellableTask) task;
+                    ((QueryGroupTask) task).setQueryGroupId(threadPool.getThreadContext());
+                    assertEquals(request.getQueryGroupId(), ((QueryGroupTask) task).getQueryGroupId());
+                    long startTime = System.nanoTime();
+
+                    while (System.nanoTime() - startTime < TIMEOUT.getNanos()) {
+                        doWork(request);
+                        if (cancellableTask.isCancelled()) {
+                            break;
+                        }
+                    }
+                    if (cancellableTask.isCancelled()) {
+                        throw new TaskCancelledException(cancellableTask.getReasonCancelled());
+                    } else {
+                        listener.onResponse(new TestResponse());
+                    }
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                }
+            });
+        }
+
+        private void doWork(TestQueryGroupTaskRequest request) throws InterruptedException {
+            switch (request.getType()) {
+                case "CPU":
+                    long i = 0, j = 1, k = 1, iterations = 1000;
+                    do {
+                        j += i;
+                        k *= j;
+                        i++;
+                    } while (i < iterations);
+                    break;
+                case "MEMORY":
+                    Byte[] bytes = new Byte[100000];
+                    int[] ints = new int[1000000];
+                    break;
```

Comment:
To make it machine agnostic we should supply the numbers in percentage of available heap bytes. Otherwise this will be flaky.